### PR TITLE
Add Machine Learning Tutorial for MusicBox

### LINKED ISSUE #363: Create Machine Learning Tutorial
This tutorial will walk the user through some very basic ML applications of MusicBox: a linear regression model and a random forest model. The system and the simulated data will be very similar to the one in the first tutorial but with a couple tweaks to simplify it.

### PR DESCRIPTION
**Pull request by aidan-winney**
Closes #363 

This pull request contains the first draft of the ML tutorial that goes over a linear regression model and a random forest model. It also contains some extremely minor changes in the other tutorials.

### CODE REVIEW

**Comment by ThomasMGeo** [comment]
Hi @aidan-winney 

1. What are the units on this RMSE? You might want to un-scale them and go for "real vs real" instead of the scaled values. 

2. When you split data, you should have the min-max scaler only fit on the input/train data. This is a form of data leakage. 

3. This is a nit, instead of shift value being a negative, I might call it t_steps_ahead and make it a positive number, and put the negative sign in the data processing part.

File: tutorial/4. machine_learning.ipynb [hunk 1]
```diff
@@ -0,0 +1,2159 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "beff6760",
+   "metadata": {},
+   "source": [
+    "## 3. Creating the Future Time Steps\n",
+    "\n",
+    "In order to build the ML model, it needs an expected output for every input for each of the species' concentrations.</br>\n",
+    "To achieve this, three new columns are made to represent the reaction one time step ahead of the current time step.</br>\n",
```
Comment (K20shores) [inline tutorial/4. machine_learning.ipynb:306]:
```suggestion
    "To achieve this, three new columns are made to represent the concentrations one time step ahead of the current time step.</br>\n",
```

File: tutorial/4. machine_learning.ipynb [hunk 2]
```diff
@@ -0,0 +1,2159 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "cc008163",
+   "metadata": {},
+   "source": [
+    "## 4. Normalizing the Concentrations\n",
+    "\n",
+    "For many ML models, including linear regression, normalizing your data will ensure that each of the input and output features are weighted equally when they are not on the same scale initially.</br>\n",
+    "This is the case for the concentrations, where X drops all the way to 4e-41 while Y never goes below 3.4, a ginormous difference in scale.</br>\n",
+    "To normalize the data, a <i>MinMaxScaler()</i> is created which can alter the concentration columns in place with its <i>fit_transform()</i> function by taking in a column that is used to calcuate the normalized values.</br>\n",
```
Comment (K20shores) [inline tutorial/4. machine_learning.ipynb:343]:
```suggestion
    "To normalize the data, a `[MinMaxScaler()](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html)` is created which can alter the concentration columns in place with its <i>fit_transform()</i> function by taking in a column that is used to calcuate the normalized values.</br>\n",
```

maybe let's link it directly

File: tutorial/4. machine_learning.ipynb [hunk 3]
```diff
@@ -0,0 +1,2159 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ea075da1",
+   "metadata": {},
+   "source": [
+    "## 5. Defining the Independent and Dependent Variables\n",
+    "\n",
+    "Next, the independent and dependent variables must be defined.</br>\n",
+    "For any ML model, your independent variables will be predicting the dependent variables</br>\n",
```
Comment (K20shores) [inline tutorial/4. machine_learning.ipynb:884]:
```suggestion
    "For any ML model, your independent variables will be predicting the dependent variables.</br>\n",
```

File: tutorial/4. machine_learning.ipynb [hunk 4]
```diff
@@ -0,0 +1,2159 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e2e0d4c0",
+   "metadata": {},
+   "source": [
+    "## 7. Fitting, Predicting, and Evaluating the Model\n",
+    "\n",
+    "With all the model preparation out of the way, it can now be created, fitted with the training data, and used to predict from the testing day.</br>\n",
+    "The with predicted y values calculated, the root mean square error (RMSE) and the correlation coefficient (R<sup>2</sup>) can be calculated.</br>\n",
```
Comment (K20shores) [inline tutorial/4. machine_learning.ipynb:944]:
```suggestion
    "With the predicted y values calculated, the root mean square error (RMSE) and the correlation coefficient (R<sup>2</sup>) can be calculated.</br>\n",
```

**Review by K20shores — CHANGES_REQUESTED** [review]
Looks good, just a few typos

File: tutorial/4. machine_learning.ipynb [hunk 5]
```diff
@@ -0,0 +1,2159 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3d9a7d13",
+   "metadata": {},
+   "source": [
+    "## 9. Plot the Linear Regression Model\n",
+    "\n",
+    "This code cell simply plots the three species' concentrations over time, with the predicted concentrations being scatterplots that represent the time steps present in the testing data.</br>\n",
+    "Due to the simplicity of this model, it is incredibly accurate, but traditional ML models will never reach this level of accuracy."
```
Comment (sjsprecious) [inline tutorial/4. machine_learning.ipynb:1506]: The example chemical reactions should have some non-linearity but I am surprised that a linear model can fit it near perfectly.

In addition, what I expect to use a ML model is that I provide the same initial condition and use the ML model to solve the same length of simulation (199 seconds in your case). Then I can compare the ML output with the ones solved numerically by MusicBox model. Will this be a better way to visualize the ML model here?

File: tutorial/2. override_mechanism.ipynb
Comment (sjsprecious) [inline tutorial/2. override_mechanism.ipynb:1]: Since you have multiple tutorials now, better to change the folder name to `tutorials`?

**Review by sjsprecious — CHANGES_REQUESTED** [review]
Thanks @aidan-winney for working on this tutorial. It looks nice to me and can be a good starting place of our ML journey.

I do not know too much about the ML details so my comments can be naive. Feel free to drop the ones that do not make sense to you.

File: tutorial/4. machine_learning.ipynb [hunk 6]
```diff
@@ -0,0 +1,2159 @@
...
+   "metadata": {},
+   "source": [
+    "## 3. Creating the Future Time Steps\n",
+    "\n",
+    "In order to build the ML model, it needs an expected output for every input for each of the species' concentrations.</br>\n",
+    "To achieve this, three new columns are made to represent the reaction one time step ahead of the current time step.</br>\n",
+    "When the model is constructed, it will attempt to predict the values one time step ahead from the values at the current time step.</br>\n",
+    "However, you can increase the number of time steps you wish to predict ahead by changing the <i>shift_value</i> to be a larger negative number.</br>\n",
+    "For example, a <i>shift_value</i> of -5 means the model will predict the concentrations five time steps ahead for every step.</br>\n",
+    "There is a fourth column also added called <i>dt</i>, which simply represents the time step length which will be fed into the model as an independent variable.</br>\n",
```
[the remark this answers is not in the document]
Comment (aidan-winney) [reply tutorial/4. machine_learning.ipynb:310]: I was thinking this for the temperature and pressure as well, but I believe the dt addition was from Thomas' comments. We can discuss further about this when you are back or I can discuss it with Kyle and Matthew.

File: tutorial/4. machine_learning.ipynb (hunk 5 above)
Comment (aidan-winney) [reply tutorial/4. machine_learning.ipynb:1506]: Yes, the accuracy of the model also puzzles me, I definitely need to discuss that.

You are right that the current plot does not visualize the full simulation since the prediction data is only from the testing data. I will see if I can get that working.

**Comment by aidan-winney** [comment]
Please let me know if there is still something wrong with the normalization procedure in step 6, especially with transforming the y columns based on the scale of the x features.

File: tutorial/4. machine_learning.ipynb (hunk 5 above)
Comment (aidan-winney) [reply tutorial/4. machine_learning.ipynb]: No, it did not.

File: tutorial/4. machine_learning.ipynb [hunk 7]
```diff
@@ -0,0 +1,2159 @@
...
+   "cell_type": "markdown",
+   "id": "cc008163",
+   "metadata": {},
+   "source": [
+    "## 4. Normalizing the Concentrations\n",
+    "\n",
+    "For many ML models, including linear regression, normalizing your data will ensure that each of the input and output features are weighted equally when they are not on the same scale initially.</br>\n",
+    "This is the case for the concentrations, where X drops all the way to 4e-41 while Y never goes below 3.4, a ginormous difference in scale.</br>\n",
+    "To normalize the data, a <i>MinMaxScaler()</i> is created which can alter the concentration columns in place with its <i>fit_transform()</i> function by taking in a column that is used to calcuate the normalized values.</br>\n",
+    "It should be noted that the t+1 columns are normalized based on the original concentration columns so that the normalization is kept consistent between the original column and the future time step column.</br>\n",
```
[the remark this answers is not in the document]
Comment (aidan-winney) [reply tutorial/4. machine_learning.ipynb]: Yes, the name was changed from "t+1" to "future" to reflect that.

File: tutorial/4. machine_learning.ipynb [hunk 8]
```diff
@@ -0,0 +1,2159 @@
...
+   "metadata": {},
+   "source": [
+    "## 5. Defining the Independent and Dependent Variables\n",
+    "\n",
+    "Next, the independent and dependent variables must be defined.</br>\n",
+    "For any ML model, your independent variables will be predicting the dependent variables</br>\n",
+    "For this model, the indepedent variables are:\n",
+    "* the temperature,\n",
+    "* the pressure,\n",
+    "* the time step length discussed in the previous step, and\n",
```
[the remark this answers is not in the document]
Comment (aidan-winney) [reply tutorial/4. machine_learning.ipynb]: Yes, I was thinking that for both dt as well as the temperature and pressure, I will need to discuss that further with others.

[the remark this answers is not in the document]
Comment (K20shores) [reply tutorial/4. machine_learning.ipynb]: They do. The rate of these reactions is determined by temperature and pressure. It's unlikely that this information helps at all because it's constant in this instance. However, that's not the point. Temperature will vary in other models, and we are only trying to show an example of how you might make a machine learning model and I think it's reasonable to give an example showing inputs that can affect chemistry, even if they are constant in this toy example

**Comment by ThomasMGeo** [comment]
@aidan-winney , apologies for the late ask, for time series, you don't want to pull randomly. 

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html

You will want a block of time at the end (or the beginning). 

https://medium.com/@mouadenna/time-series-splitting-techniques-ensuring-accurate-model-validation-5a3146db3088

Right now your predicting Friday's high temp, but you have Thursday and Saturday's value known.

File: tutorial/4. machine_learning.ipynb (hunk 8 above)
[the remark this answers is not in the document]
Comment (sjsprecious) [reply tutorial/4. machine_learning.ipynb]: Sounds good and thanks for your clarification.

**Comment by ThomasMGeo** [comment]
You will want ~70% of your timeseries for training, and ~30% for testing, and they should not overlap at all.

**Comment by aidan-winney** [comment]
The TimeSeriesSplit class seems useful, but it introduces k-fold cross-validation which I do not think is necessary for this tutorial. Cross-validation will blow up the number of plots from 2 to 2 * num_folds, where 5 is a commonly used value there, leading to 10 plots. Instead, do you think simply turning shuffle off for the train_test_split() function works? The default behavior is to take the back end as testing data, which is what is desired here.

**Comment by aidan-winney** [comment]
Here is an image of turning shuffle off for reference (still suspicious on the level of accuracy this has, let me know if you have any idea as to why a linear model is fitting so well).

image

**Comment by ThomasMGeo** [comment]
Yeah you don't have to use that specific function, can easily roll your own to split the time series. Just want the time series to be in order and split up if that makes sense? Happy to chat next week.

**Comment by aidan-winney** [comment]
I had to fix some code because I realized the visualization was not considering un-shifting the future time back to the time steps it is actually predicting. However, this did not change the accuracy of the predictions at all; it is perfect even when shifted 100+ time steps ahead, which is very confusing. I have been unsuccessful in finding where the data leakage is originating from. However, this change I made completely broke the random forest model, leading to it predicting constant concentrations, which is obviously incorrect. Not sure what is going on at this point.

**Review by K20shores — APPROVED** [review]
